testcases/kernel/syscalls/madvise/madvise07.c

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) 2016-2017 Richard Palethorpe <richiejp@f-m.fm>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ * Check that accessing memory marked with MADV_HWPOISON results in SIGBUS.
+ *
+ * Test flow: map memory, 
+ *	      create child process, 
+ *	      mark memory with MADV_HWPOISON inside child process,
+ *	      access memory,
+ *	      if SIGBUS is delivered to child the test passes else it fails
+ */
+
+#include "tst_test.h"
+#include <fcntl.h>
+#include <unistd.h>
+#include <signal.h>
+
+#ifdef MADV_HWPOISON
+
+struct test_case {
+	int fd;
+	char *fname;
+	char *fcontent;
+	void *fmem;
+	struct stat stat;
+};
+static struct test_case cases[1];
+
+static void setup(void)
+{
+	cases[0].fd = 0;
+	cases[0].fcontent = "abcdefghijklmnopqrstuvwxyz12345\n";
+	cases[0].fname = "mmf0";
+	cases[0].fmem = NULL;
+}
+
+static void cleanup(void)
+{
+	unsigned int i;
+	struct test_case c;
+
+	for (i = 0; i < ARRAY_SIZE(cases); i++) {
+		c = cases[i];
+		if (c.fmem != NULL && munmap(c.fmem, c.stat.st_size))
+			tst_brk(TBROK | TERRNO, "munmap(cases[%d].fmem)", i);
+
+		c.fmem = NULL;
+		if (c.fd > 0 && close(c.fd))
+			tst_brk(TBROK | TERRNO, "close(cases[%d].fd)", i);
+	}
+}
+
+static void on_sigbus(int s __attribute__((__unused__)))
+{
+	tst_res(TPASS, "Received SIGBUS from madvise with MADV_HWPOISON");
+	_exit(0);
+}
+
```

Comment:
Calling tst_res() from signal handler is not 100% safe. It calls functions that operate on FILE to print error messages which may allocate memory, etc.

More elegant solution would be:

* just dereference the memory in child
* check that it was terminated by a signal in parent
  (that is what waitpid() is for)